File: cmrdv_ws/src/cmrdv_common/cmrdv_common/heartbeat/heart.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String
from cmrdv_interfaces.msg import Heartbeat
from cmrdv_ws.src.cmrdv_common.cmrdv_common.config.collection_config import RELIABLE_QOS_PROFILE
from cmrdv_ws.src.cmrdv_common.cmrdv_common.config.common_config import GLOBAL_TIMEOUT_SEC
from cmrdv_ws.src.cmrdv_common.cmrdv_common.CAN.can_types import ALIVE_STATE, WARNING_STATE, ERROR_STATE, TX10HZ_PERIOD_S

# from cmrdv_common.config.collection_config import RELIABLE_QOS_PROFILE
# from cmrdv_common.config.common_config import GLOBAL_TIMEOUT_SEC
# from cmrdv_common.CAN.can_types import ALIVE_STATE, WARNING_STATE, ERROR_STATE, TX10HZ_PERIOD_S


class HeartbeatNode(Node):
    """
    Base class for each vital node to inherit. Creates a publisher for a /`{node}_status` topic and 
    subscribes to the /global_heartbeat topic. Default timer period is 10hz for updating the nodes status
    and reading the global status

    Parameters
    ----------
    node_name : str
        Node name found on the topic, `{node}_status`
    timer_period : float
        Frequency to publish status to system
    """
    def __init__(self, node_name, timer_period=TX10HZ_PERIOD_S):
        super().__init__(node_name)
        self.node_name = node_name

        self.time = self.get_clock().now()
        self.timer = self.create_timer(timer_period, self.node_status_pub)

        self.threshold = rclpy.time.Duration(seconds=(4 * timer_period))

        # TODO: perhaps change node status to correspond to CAN status
        self.node_status = ALIVE_STATE
        self.global_status = ALIVE_STATE

        self.status_publisher = self.create_publisher(
            Heartbeat,
            f"{node_name}_status",
            RELIABLE_QOS_PROFILE
        )
        self.global_heartbeat = self.create_subscription(
            Heartbeat, 
            "global_heartbeat", 
            self.update_status, 
            RELIABLE_QOS_PROFILE
        )

        # TODO: why use global timeout sec and not check at 10hz?
        self.last_global_heartbeat = self.get_clock().now()
        self.check_global_timeout = self.create_timer(
            GLOBAL_TIMEOUT_SEC, 
            self.check_global_heartbeat_timeout
        )
        self.global_heartbeat  # prevent unused variable warning

    def node_status_pub(self):
        self.time = self.get_clock().now()

        heartbeat = Heartbeat()

        # NOTE: udpating the node status to be in line with the global status
        if (self.global_status != ERROR_STATE and self.node_status != ERROR_STATE): 
            node_status = self.global_status
        else: # can add more logic here
            node_status = ERROR_STATE

        if node_status != self.node_status: 
            self.get_logger().info(f"publishing {self.node_name} status: {self.node_status}")

        self.node_status = node_status

        heartbeat.status = self.node_status
        heartbeat.header.stamp = self.time.to_msg()
        heartbeat.header.frame_id = self.node_name

        self.status_publisher.publish(heartbeat)

    def check_global_heartbeat_timeout(self):
        if self.get_clock().now() - self.last_global_heartbeat > self.threshold:
            self.global_status = ERROR_STATE
            self.get_logger().info(f'GLOBAL HEARTBEAT TIMEOUT!')

    def update_status(self, msg):
        self.global_status = msg.status
        self.last_global_heartbeat = self.get_clock().now()
        # self.get_logger().info(f"receving hb status: {msg.status}")

    def panic(self):
        self.node_status = ERROR_STATE
    
    def clear_error(self):
        self.node_status = ALIVE_STATE

    def alive(self):
        return (self.node_status != ERROR_STATE)


class Perceptions(HeartbeatNode):
    def __init__(self):
        super().__init__('perceptions')
        self.count = 0  
        self.timer = self.create_timer(0.5, self.increment_count)

    def increment_count(self):
        self.count += 1
        print(self.count)
        if not self.alive(): 
            print("ERROR")
        if self.count > 10: 
            self.panic()

class Planning(HeartbeatNode):
    def __init__(self):
        super().__init__('planning')
        self.count = 0  
        self.timer = self.create_timer(1, self.increment_count)

    def increment_count(self):
        self.count += 1
        if not self.alive(): 
            print("ERROR")
        if self.count > 100: 
            self.panic()

class DIM(HeartbeatNode):
    def __init__(self):
        super().__init__('dim')
        self.count = 0  
        self.timer = self.create_timer(1, self.increment_count)
        self.state_subscriber = self.create_subscription(
                String,
                "DIM_request",
                self.process_state_request,
                10
        )

    def process_state_request(self, msg):
        self.get_logger().info(f"DIM REQUEST: {msg.data}")

    def increment_count(self):
        self.count += 1
        if not self.alive(): 
            print("ERROR")

def dim(args=None):
    rclpy.init(args=args)

    dim = DIM()

    rclpy.spin(dim)

    dim.destroy_node()
    rclpy.shutdown()

def perceptions(args=None):
    rclpy.init(args=args)

    perceptions = Perceptions()

    rclpy.spin(perceptions)

    perceptions.destroy_node()
    rclpy.shutdown()

def planning(args=None):
    rclpy.init(args=args)

    perceptions = Planning()

    rclpy.spin(perceptions)

    perceptions.destroy_node()
    rclpy.shutdown()

def main(args=None):
    rclpy.init(args=args)

    node = Heartbeat("node")

    rclpy.spin(node)

    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: driverless_ws/src/heartbeat/src/heart.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rcllscpp.hpp"
#include "std_msgs/msg/string.hpp"

// #include "interfaces/msg/Heartbeat.hpp"

// #include "common/config/collection_config.hpp"
// #include "common/config/common_config.hpp"
// #include "common/CAN/can_types.hpp"

// Usage of the imported constants/enums in your C++ code
const auto RELIABLE_QOS_PROFILE = common::config::RELIABLE_QOS_PROFILE;
const auto GLOBAL_TIMEOUT_SEC = common::config::GLOBAL_TIMEOUT_SEC;
const auto ALIVE_STATE = common::CAN::ALIVE_STATE;
const auto WARNING_STATE = common::CAN::WARNING_STATE;
const auto ERROR_STATE = common::CAN::ERROR_STATE;
const auto TX10HZ_PERIOD_S = common::CAN::TX10HZ_PERIOD_S;


// don't forget to import types when done

class HeartbeatNode : public rclcpp::Node {
  public:
    HeartbeatNode(const std::string& node_name, double timer_period = TX10HZ_PERIOD_S) 
    : Node(node_name),
      node_name_(node_name)
    {
      time_ = rclcpp::Clock().now();
      timer_ = this->create_wall_timer(std::chrono::duration<double>(timer_period_), 
                                std::bind(&HeartbeatNode::node_status_pub, this));

      threshold_ =rclcpp::Duration(4 * timer_period);

      // TODO: perhaps change node status to correspond to CAN status
      node_status_ = ALIVE_STATE;
      global_status_ = ALIVE_STATE;

      status_publisher_ = create_publisher<interfaces::msg::Heartbeat>(node_name + "_status", rclcpp::QoS(10).reliable());
      global_heartbeat_subscription_ = create_subscription<interfaces::msg::Heartbeat>(
          "global_heartbeat",
          rclcpp::QoS(10).reliable(),
          std::bind(&HeartbeatNode::update_status, this, std::placeholders::_1));

			// TODO: why use global timeout sec and not check at 10hz?
			last_global_heartbeat_ = rclcpp::Clock().now();
			check_global_timeout_ = this->create_wall_timer(std::chrono::seconds(GLOBAL_TIMEOUT_SEC),
																					std::bind(&HeartbeatNode::check_global_heartbeat_timeout, this));
    }


		void node_status_pub() {
			time_ = rclcpp::Clock().now();

			interfaces::msg::Heartbeat heartbeat;

			// NOTE: Updating the node status to be in line with the global status
			// what is node_status??????
			if (global_status_ != ERROR_STATE && node_status_ != ERROR_STATE) {
					node_status = global_status_;
			} else {
					// You can add more logic here if needed
					node_status = ERROR_STATE;
			}

			if (node_status != node_status_) {
					RCLCPP_INFO(get_logger(), "publishing %s status: %s", node_name_.c_str(), node_status.c_str());
			}

			node_status_ = node_status;

			heartbeat.status = node_status_;
			heartbeat.header.stamp = time.to_msg();
			heartbeat.header.frame_id = node_name_;

			status_publisher_->publish(heartbeat);
		}


		void check_global_heartbeat_timeout() {
			rclcpp::Time current_time = rclcpp::Clock().now();

			if (current_time - last_global_heartbeat_ > threshold_) {
					global_status_ = ERROR_STATE;
					RCLCPP_INFO(get_logger(), "GLOBAL HEARTBEAT TIMEOUT!");
			}
		}

		void update_status(const interfaces::msg::Heartbeat::SharedPtr msg) {
			global_status_ = msg->status;
			last_global_heartbeat_ = rclcpp::Clock().now();
			// get_logger()->info("receving hb status: " + std::to_string(msg->status));
		}

		void panic() {
			node_status_ = ERROR_STATE;
		}

		void clear_error() {
			node_status_ = ALIVE_STATE;
		}

		bool alive() {
			return (node_status_ != ERROR_STATE);
		}

}


class Perceptions : public HeartbeatNode {
	public:
		Perceptions() : HeartbeatNode("perceptions") {
			count_ = 0;
			timer_ = this->create_wall_timer(std::chrono::milliseconds(500), 
															std::bind(&Perceptions::increment_count, this));
		}

	private:
		void increment_count() {
			count_++;
			std::cout << count_ << std::endl;
			if (!alive()) { // how to call parent function?
				std::cout << "ERROR" << std::endl;
			}
			if (count_ > 10) {
				panic();
			}
		}

		int count_;
  	rclcpp::TimerBase::SharedPtr timer_;

};




class Planning : public HeartbeatNode {
	public:
		Planning() : HeartbeatNode("planning") {
			count_ = 0;
			timer_ = this->create_wall_timer(std::chrono::seconds(1), 
																std::bind(&Planning::increment_count, this));
		}

	private:
		void increment_count() {
			count_++;
			if (!alive()) {
				std::cout << "ERROR" << std::endl;
			}
			if (count_ > 100) {
				panic();
			}
		}

		int count_;
    rclcpp::TimerBase::SharedPtr timer_;

};


class DIM : public HeartbeatNode {
	public:
		DIM() : HeartbeatNode("dim") {
			count_ = 0;
			timer_ = this->create_wall_timer(std::chrono::seconds(1), [this]() {
				increment_count();
			});

			state_subscriber_ = this->create_subscription<std_msgs::msg::String>(
				"DIM_request", 10,
				std::bind(&DIM::process_state_request, this, std::placeholders::_1)
			);
		}

	private:
		void process_state_request(const std_msgs::msg::String::SharedPtr msg) {
			RCLCPP_INFO(this->get_logger(), "DIM REQUEST: %s", msg->data.c_str());
		}

		void increment_count() {
			count_++;
			if (!alive()) {
				std::cout << "ERROR" << std::endl;
			}
		}

		int count_;
		rclcpp::TimerBase::SharedPtr timer_;
		rclcpp::Subscription<std_msgs::msg::String>::SharedPtr state_subscriber_;
	};



int main(int argc, char **argv) {
	rclcpp::init(argc, argv);

	// DIM Node
	// auto dimNode = std::make_shared<DIM>();
	// rclcpp::spin(dimNode);
	// rclcpp::shutdown();

	// Perceptions Node
	// auto perceptionsNode = std::make_shared<Perceptions>();
	// rclcpp::spin(perceptionsNode);
	// rclcpp::shutdown();

	// Planning Node
	// auto planningNode = std::make_shared<Planning>();

	auto node = std::make_shared<HeartbeatNode>("node");

	rclcpp::spin(planningNode);
	rclcpp::shutdown();

	return 0;
}